src/erlcloud_ec2.erl

```diff
@@ -3574,6 +3577,142 @@ ec2_query(Config, Action, Params, ApiVersion) ->
+    flatten(map_to_kv(ParamsMap));
+prepare_action_params(ParamsMap, Filters) when is_map(ParamsMap) ->
+    % Convert to list of tuples, with {Key, Value} format for each record
+    % Flatten the list of tuples
+    FlatList = flatten(map_to_kv(ParamsMap)),
+    FlatList ++ list_to_ec2_filter(Filters). % Add the filters 
+
+map_to_kv(Map) when is_map(Map) ->
+    MapList = maps:fold(fun
+        (Key, Value, Acc) ->
+            [{Key, map_to_kv(Value)} | Acc]
+    end, [], Map),
+    lists:reverse(MapList);
+map_to_kv(Val) ->
+    Val.
+
+flatten(Proplist) ->
+    lists:flatten(flatten_recursive(Proplist, <<>>)).
+
+flatten_recursive([], _ParentKey) ->
+    [];
+flatten_recursive([{Key, Val} | Rest], ParentKey) ->
+    case {is_list_of_kv_tuples(Val), is_list_not_string(Val)} of
+        {true,_} -> % If Val is a nested record
+            [flatten_recursive(Val, concat_key(ParentKey, Key)) | flatten_recursive(Rest, ParentKey)];
+        {_,true} -> % Is a nested list
+            [generate_kv_tuples(concat_key(ParentKey, Key), Val) | flatten_recursive(Rest, ParentKey)];
+        _ -> % If Val is a non-nested value - Such as a string
+            [{concat_key(ParentKey, Key), Val} | flatten_recursive(Rest, ParentKey)]
+    end.
+
+generate_kv_tuples(Key, Values) ->
+    generate_kv_tuples(Key, Values, 1, []).
+
+generate_kv_tuples(_, [], _, Acc) ->
+    lists:reverse(Acc);
+generate_kv_tuples(Key, [Value | Rest], Index, Acc) ->
+    NewKey = concat_key(Key, integer_to_list(Index)),
+    generate_kv_tuples(Key, Rest, Index + 1, [{NewKey, Value} | Acc]).    
+
+concat_key(<<>>, Key) when is_bitstring(Key); is_atom(Key)  ->
+    to_bitstring(underscore_to_pascal_case(Key));
+concat_key(<<>>, Key) when is_list(Key) -> 
+    Key;
+concat_key(ParentKey, Key) ->
+    BiParentKey = to_bitstring(underscore_to_pascal_case(ParentKey)),
+    BiKey = to_bitstring(underscore_to_pascal_case(Key)),
+    <<BiParentKey/binary, ".", BiKey/binary>>.
+
+is_string(Val) when is_list(Val)-> 
+    io_lib:printable_list(Val) or io_lib:printable_list(Val);
+is_string(_) ->
+    false.
+ 
+is_list_not_string(Val) when is_list(Val)-> 
+    not is_string(Val);
+is_list_not_string(_) ->
+    false.
+
+is_list_of_kv_tuples(List) when is_list(List) ->
+    lists:all(fun is_tuple/1, List) andalso lists:all(fun is_kv_tuple/1, List);
+is_list_of_kv_tuples(_) ->
+    false.
+
+is_kv_tuple({_, _}) -> true;
+is_kv_tuple(_) -> false.
+
+%%% Conversions
+to_bitstring(In) when is_bitstring(In) ->
+    In;
+to_bitstring(In) when is_list(In) ->
+    list_to_binary(In);
+to_bitstring(In) when is_atom(In) ->
+    erlang:atom_to_binary(In).
+
+capitalize_word([First | Rest]) ->
+    FirstUpper = list_to_uppercase([First]),
+    FirstUpper ++ Rest.
+
+list_to_uppercase(List) ->
+    lists:map(fun(Char) -> 
+        if Char >= $a, Char =< $z -> Char - 32;
+           true -> Char
+        end
+    end, List).
+
+underscore_to_pascal_case(String) when is_list(String) ->
```

Comment:
can't we just use the same pascal_case in parameters without transformation?

Reply:
The transformation is partly left over from when we were using records - but I have left it in currently as it allows the client to specify keys in either 'PascalCase' or 'snake_case'. 

If we want we can drop support for snake_case

Reply:
I would try to keep the number of transformations to a minimum